viewer/mlab-view

```diff
@@ -0,0 +1,176 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Summary:
+  mlab-view provides a limited web server to view the collectd-web CGI scripts.
+
+  While mlab-view is running, an operator can only access the CGI scripts using
+  an SSH tunnel. Direct web connections are refused with "403 Forbidden."
+
+  mlab-view should only be used for debugging and node triage. When a node is
+  in normal operation, mlab-view should not be running. mlab-view guarantees
+  that the web server is shutdown when the operator's ssh connection has
+  disconnected.
+
+Example usage (from operator's workstation):
+
+  $ ssh -L 8088:utility.mlab.<hostname>:8088 \\
+      -p806 mlab_utility@<hostname> sudo mlab-view
+
+  Operator directs their browser to:
+
+      http://localhost:8088/
+
+What's happening?
+
+ * SSH binds to port 8088 on localhost.
+ * SSH login to <hostname> as the mlab_utility user and runs the command:
+
+     "sudo mlab-view".
+
+ * mlab-view starts a web server that listens on port 8088.
+ * The web server is configured to reject connections from any address other
+   than the slice IP.
+ * When ssh detects a connection to localhost:8088, the remote end attempts a
+   connection to:
+
+      "utility.mlab.<hostname>:8088"
+
+   To the server, this connection appears to come from the slice IP.
+ * SSH continues to forward all traffic back and forth from localhost:8088
+   through the ssh connection to the remote web server.
+"""
+
+import os
+import signal
+import subprocess
+import sys
+import syslog
+import time
+
+# Third-party modules.
+import netifaces
+
+
+ACCESS_CONF = '/usr/share/collectd-mlab/access.conf'
+HTTPD_PID = '/etc/httpd/run/httpd.pid'
+
+def log(msg):
+  """Sends log message to syslog."""
+  print msg
+  syslog.syslog(syslog.LOG_ERR, 'mlab-view: %s' % msg)
+
+
+def ignore_signal_handler(signal, frame):
+  """Handler to ignore any signal."""
+  log('Ignoring signal: %s' % signal)
+
+
+def get_address():
+  """Returns local IPv4 & IPv6 addresses as a string separated by spaces."""
+  ifaces = netifaces.interfaces()
+  addrs = []
+  for iface in ifaces:
+    if iface == 'lo':
+      continue
+    addresses = netifaces.ifaddresses(iface)
+    for addr_type in (netifaces.AF_INET, netifaces.AF_INET6):
+      if addr_type in addresses:
+        addrs.extend([addr['addr'] for addr in addresses[addr_type]])
+  if not addrs:
+    return '127.0.0.1'
+  return ' '.join(addrs)
+
+
+def get_access_config():
+  return """
+<Directory /usr/share/collectd/>
+    AddHandler cgi-script .cgi
+    DirectoryIndex bin/index.cgi
+    Order Deny,Allow
+    Allow from {ips}
+    Deny from all
+</Directory>
+""".format(ips=get_address())
+
+
+def find_last_parent(pid):
+  """Walks up process tree starting at pid, stopping at sshd or ppid == 1."""
```

Comment:
Can we clarify this function a bit? I think "last parent" is kind of ambiguous because it might refer to either to last as in "oldest" or last as in "most recent." Also it would be good to state explicitly what we're returning.